codalab/client/json_api_client.py

```diff
@@ -0,0 +1,299 @@
+    else:
+        packed_objects = pack_object(objects)
+
+    return {
+        'data': packed_objects
+    }
+
+
+class JsonApiRelationship(object):
+    """
+    Placeholder for a relationship to another resource.
+    Used to build requests to create or update a resource.
+
+    For example, to update the owner of a bundle:
+        client.update('bundles', uuid, data={
+            'owner': JsonApiRelationship('users', owner_id)
+        })
+    """
+    def __init__(self, type_, id_):
+        self.type_ = type_
+        self.id_ = id_
+
+
+class JsonApiClient(RestClient, BundleClient):
+    """
+    Simple JSON API client.
+    """
+    def __init__(self, address, auth_handler, authenticate):
+        self.address = address
+        self.auth_handler = auth_handler
+        self._authenticate = authenticate
+
+        base_url = address + '/rest'
+        super(JsonApiClient, self).__init__(base_url)
+
+    def _get_access_token(self):
+        return self._authenticate(self)
+
+    @staticmethod
+    def _get_resource_path(resource_type, resource_id=None):
+        if resource_id is None:
+            return '/' + resource_type
+        else:
+            return '/' + resource_type + '/' + str(resource_id)
+
+    @staticmethod
+    def _pack_params(params):
+        """
+        Process lists into comma-separated strings, and booleans into 1/0.
+        """
+        if params is None:
+            return None
+
+        result = {}
+        for k, v in (params.iteritems() if isinstance(params, dict) else params):
+            if isinstance(v, list):
+                result[k] = ','.join(map(unicode, v))
+            elif isinstance(v, bool):
+                result[k] = int(v)
+            else:
+                result[k] = v
+        return result
+
+    @wrap_exception
+    def fetch(self, resource_type, resource_id=None, params=None):
+        url = self._get_resource_path(resource_type, resource_id)
+        params = self._pack_params(params)
+        return unpack_document(
+            self._make_request('GET', url, query_params=params))
+
+    @wrap_exception
+    def create(self, resource_type, data, params=None):
+        url = self._get_resource_path(resource_type)
+        params = self._pack_params(params)
+        data = pack_document(data, resource_type)
+        return unpack_document(
+            self._make_request('POST', url, query_params=params, data=data))
+
+    @wrap_exception
+    def update(self, resource_type, data, params=None):
+        """
+        Request to update a resource or resources.
+
+        :param resource_type: resource type as string
+        :param data: update dict or list of update dicts, update dicts
+        must contain 'id' fields specifying the resources to update
+        :param params: dict of query parameters
+        :return:
+        """
+        if isinstance(data, list):
+            path = self._get_resource_path(resource_type)
```

Comment:
`cl chown` updates the owner of potentially multiple bundles, and so this allows a batch update. But as you pointed out with delete, I can simplify this down to always assuming batch operations.